gateway/ln-gateway/src/state_machine/pay.rs

```diff
@@ -404,36 +422,38 @@ impl GatewayPayInvoice {
                 )
                 .await
                 {
-                    return Self::buy_preimage_via_direct_swap(
+                    Self::buy_preimage_via_direct_swap(
                         client,
                         payment_parameters.invoice.clone(),
                         contract.clone(),
                         common.clone(),
                     )
-                    .await;
+                    .await
+                } else {
+                    Self::buy_preimage_over_lightning(
+                        context,
+                        payment_parameters,
+                        contract.clone(),
+                        common.clone(),
+                    )
+                    .await
+                }
+            }
+            Err(e) => {
+                warn!("Failed to get payment parameters: {e:?}");
+                match e.contract.clone() {
+                    Some(contract) => GatewayPayStateMachine {
+                        common,
+                        state: GatewayPayStates::CancelContract(Box::new(
+                            GatewayPayCancelContract { contract, error: e },
+                        )),
+                    },
+                    None => GatewayPayStateMachine {
+                        common,
+                        state: GatewayPayStates::OfferDoesNotExist(e.contract_id),
+                    },
```

Comment:
Change in behavior: We now allow cancellation of contracts if an internal payment fails (which it never should), seems ok.